purchases/src/test/java/com/revenuecat/purchases/PurchasesTest.kt

```diff
@@ -3290,6 +3301,96 @@ class PurchasesTest {
         assertThat(Purchases.sharedInstance.appConfig.finishTransactions).isTrue()
     }
 
+    @Test
+    fun `Send error if cannot find the old purchase associated when upgrading a SKU`() {
+        setup()
+
+        val activity: Activity = mockk()
+        val (skuDetails, offerings) = stubOfferings("onemonth_freetrial")
+
+        val oldPurchase = mockk<PurchaseHistoryRecordWrapper>()
+        every { oldPurchase.sku } returns "oldSku"
+        every { oldPurchase.type } returns PurchaseType.SUBS
+
+        every {
+            mockBillingWrapper.findPurchaseInPurchaseHistory(PurchaseType.SUBS.toSKUType()!!, "oldSku", captureLambda())
+        } answers {
+            lambda<(BillingResult, PurchaseHistoryRecordWrapper?) -> Unit>().captured.invoke(BillingResult(), null)
+        }
+
+        var receivedError: PurchasesError? = null
+        var receivedUserCancelled: Boolean? = null
+        purchases.purchasePackageWith(
+            activity,
+            offerings[stubOfferingIdentifier]!!.monthly!!,
+            UpgradeInfo(oldPurchase.sku),
+            onError = { error, userCancelled ->
+                receivedError = error
+                receivedUserCancelled = userCancelled
+            },
+            onSuccess =  { _, _ -> }
+        )
+
+        verify (exactly = 0) {
+            mockBillingWrapper.makePurchaseAsync(
+                eq(activity),
+                eq(appUserId),
+                skuDetails,
+                ReplaceSkuInfo(oldPurchase),
+                stubOfferingIdentifier
+            )
+        }
+        assertThat(receivedError).isNotNull
+        assertThat(receivedError!!.code).isEqualTo(PurchasesErrorCode.PurchaseInvalidError)
+        assertThat(receivedUserCancelled).isFalse()
```

Comment:
same question here, are these guaranteed to happen in order?

Reply:
I hope the other explanation explains it, but yes. In this case, we are doing the mocking and capturing in line 3318. The completion block of `findPurchaseInPurchaseHistory` will be called with `BillingResult()` and `null`.